### PR TITLE
Expose Engine.Searcher provider to ingest plugins.

### PR DESCRIPTION
**Pull request by martijnvg**
Note that this is a PR against the enrich branch and will be backported to enrich-7.x branch.

Needed for #32789 in order for ingest processors to perform local Lucene searches.

### CODE REVIEW

**Comment by martijnvg** [comment]
Pinging @elastic/es-core-infra to raise awareness of this change. This PR modifies the ingest plugin extension to provide the ability to an ingest processor to access a locally allocated shard.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 1]
```diff
@@ -96,7 +99,19 @@ public IngestService(ClusterService clusterService, ThreadPool threadPool,
                 threadPool.getThreadContext(), threadPool::relativeTimeInMillis,
                 (delay, command) -> threadPool.schedule(
                     command, TimeValue.timeValueMillis(delay), ThreadPool.Names.GENERIC
-                ), this
+                ), this, index -> {
+                    IndexService indexService = indicesService.indexService(index);
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java]: I assume from the exception message that the IndexService returns null unless it it local to this node ? (it is not very clear from the code if InicesService is per node or per cluster).

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 2]
```diff
@@ -96,7 +99,19 @@ public IngestService(ClusterService clusterService, ThreadPool threadPool,
                 threadPool.getThreadContext(), threadPool::relativeTimeInMillis,
                 (delay, command) -> threadPool.schedule(
                     command, TimeValue.timeValueMillis(delay), ThreadPool.Names.GENERIC
-                ), this
+                ), this, index -> {
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java]: I think we will need to do some alias resolution here as it appears that this call requires the concrete index ?

Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: I see now you do it in the test processor ... but maybe move the logic here to have one place for failures of that kind ?

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 3]
```diff
@@ -96,7 +99,19 @@ public IngestService(ClusterService clusterService, ThreadPool threadPool,
...
+                    if (indexService == null) {
+                        throw new ResourceNotFoundException("index [" + index + "] not locally allocated");
+                    }
+                    int numShards = indexService.getMetaData().getNumberOfShards();
+                    if (numShards != 1) {
+                        throw new IllegalStateException("index [" + index + "] must have 1 shard, but has " + numShards + " shards");
+                    }
+
+                    IndexShard indexShard = indexService.getShard(0);
+                    return indexShard.acquireSearcher("ingest");
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java]: for my education, does this create a new searcher or give you back a pre-existing one ? If the former, do we need to do anything with the existing one w/r/t disabling refreshes ? (sorry if this a bogus question, I don't have much experience at this level).

File: server/src/main/java/org/elasticsearch/ingest/Processor.java [hunk 4]
```diff
@@ -110,16 +113,24 @@ Processor create(Map<String, Processor.Factory> processorFactories, String tag,
...
+         * Provides access to an engine searcher of a locally allocated index specified for the provided index.
+         *
+         * The locally allocated index must be have a single primary shard.
+         */
+        public final Function<Index, Engine.Searcher> searcherProvider;
+
         public Parameters(Environment env, ScriptService scriptService, AnalysisRegistry analysisRegistry,  ThreadContext threadContext,
                           LongSupplier relativeTimeSupplier, BiFunction<Long, Runnable, Scheduler.ScheduledCancellable> scheduler,
-            IngestService ingestService) {
+                          IngestService ingestService, Function<Index, Engine.Searcher> searcherProvider) {
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/ingest/Processor.java]: nit: searcherProvider -> localShardSearcher

**Comment by jakelandis** [comment]
LGTM, just a couple questions and a nitpick

However, I am a bit concerned by exposing this to all ingest plugins and would like another opinion on that point before fully commiting to this approach.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 2 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: :+1: yeah it makes sense to move the index expression resolution to here.

The `Function<Index, Engine.Searcher>` would then change into `Function<String, Engine.Searcher>`.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 3 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: This always creates a new searcher. The code invoking this should be aware of this and make sure to close it after it is done with the searcher. In our case we will need to reuse the searcher from an ingest processor. We don't want to invoke this every time an enrich processor is invoked.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: Yes, `IndicesServices` only contains a `IndexService` instance for locally allocated indices.

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by martijnvg** [comment]
> However, I am a bit concerned by exposing this to all ingest plugins and would like another opinion on that point before fully commiting to this approach.

Not sure how else we can expose a local allocated shard to an ingest processor. We either allow it for all plugins or not at all. I think the same applies also to other components (ClusterService, ScriptService etc.) which we already expose to ingest plugins.

**Comment by jakelandis** [comment]
I am not advocating for nor against exposing this as-is. I simply don't fully understand the ramifications (if any) this may brings. 

@jpountz or @jasontedor  - do you see any potential issues of exposing an IndexSearcher (for local single shard index) directly to any Ingest plugin ? 


> Not sure how else we can expose a local allocated shard to an ingest processor.

There is likely a way, but would require some refactoring.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 5]
```diff
@@ -86,17 +92,38 @@
...
                 threadPool.getThreadContext(), threadPool::relativeTimeInMillis,
                 (delay, command) -> threadPool.schedule(
                     command, TimeValue.timeValueMillis(delay), ThreadPool.Names.GENERIC
-                ), this
+                ), this, indexExpression -> {
+                    ClusterState state = clusterService.state();
+                    Index[] resolvedIndices = resolver.concreteIndices(state, IndicesOptions.STRICT_EXPAND_OPEN, indexExpression);
+                    if (resolvedIndices.length != 1) {
+                        throw new IllegalStateException("expression [" + indexExpression + "] can only point to a single concrete index");
+                    }
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java]: should we prevent filtered aliases too?

**Review by jpountz — COMMENTED** [review]
> @jpountz or @jasontedor - do you see any potential issues of exposing an IndexSearcher (for local single shard index) directly to any Ingest plugin ?

The main issue I can think of is that the searcher might provide access to more information than it should, eg. if a filtered alias is queried, if the index is not allowed for reading, or if there are FLS/DLS rules configured. Searchers are read-only so there is no risk of altering data however.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 5 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: Yes, that makes sense. There is no guarantee that they will be applied in an ingest processor. I will fail with an error if an alias has a filter.

**Comment by martijnvg** [comment]
> or if there are FLS/DLS rules configured

I think FLS/DLS rules still apply, because `IndexShard#acquireSearcher(...)` uses an `IndexSearcherWrapper` it has been configured and that is the case if dls/fls has been configured.

**Comment by jpountz** [comment]
Let's have a test that the searcher wrapper is applied when acquired via an ingest processor?

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

File: server/src/test/java/org/elasticsearch/ingest/IngestLocalShardSearcherTests.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+        TestProcessor(String tag, Function<String, Engine.Searcher> localShardSearcher) {
+            super(tag);
+            this.localShardSearcher = localShardSearcher;
+        }
+
+        @Override
+        public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
+            String indexExpression = "reference-index";
+            try (Engine.Searcher engineSearcher = localShardSearcher.apply(indexExpression)) {
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/ingest/IngestLocalShardSearcherTests.java]: Asserting that the wrapper has been called doesn't feel enough to me, I'd like to have a custom directory reader that the wrapper uses, and assert here that we get the expected implementation.

Comment (martijnvg) [reply server/src/test/java/org/elasticsearch/ingest/IngestLocalShardSearcherTests.java]: I've adjusted this test: https://github.com/elastic/elasticsearch/pull/41010/commits/fc939ef1b37725dfc9b9fa23d8559364539d0057

**Review by jpountz — APPROVED** [review]
Thanks @martijnvg !

**Review by jakelandis — APPROVED** [review]

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/1

(unrelated test failure, IntervalQueryBuilderTests.testToQuery, java.lang.IllegalArgumentException: Too many disjunctions to expand)

**Pull request merged**